### ISSUE TITLE
Git: stale editor inputs hanging around

### ISSUE THREAD

**Issue by bpasero** [microsoft/vscode#18651]
I am seeing multiple editors in index and working tree hanging around even though those are no longer valid:

![image](https://cloud.githubusercontent.com/assets/900690/22016900/40dd1b9a-dca8-11e6-8cfa-3a620992f7df.png)

**Comment by joaomoreno** [comment]
There's no control of this from an extension side. The current API has:

- `provideContent(): string`
- `onDidChange(): Event<Uri>`

What I would need is an `onDidDispose` event. I need a way to tell someone that a specific resource I once provided content for no longer exists. We could then do the right thing throughout the workbench.

Thoughts @jrieken @bpasero ?

**Comment by jrieken** [comment]
Unsure what to do there. The API just talks about content provider so I am unsure if they should be allows to define the lifecycle of the document...

Maybe the issue is that we don't have a nice dispose story for virtual documents. @joaomoreno Would it help if we start to dispose virtual document more clever? So start to ref count them and instead of the 3mins rule we could say virtual documents get disposed when they aren't 'used' anymore - where used means on the extension host not referenced anymore and on the main side not visible anymore.

**Comment by joaomoreno** [comment]
That does work if the extension also participates in the ref counting. 👍

**Comment by jrieken** [comment]
I can investigate...

**Comment by bpasero** [comment]
@joaomoreno @jrieken did you guys figure out a solution? I think today the experience is pretty bad with diff editors piling up from working tree and index even though I already committed my stuff. I think we need an explicit way for extension (or any document provider) to indicate when a resource is no longer valid. I would need the same once `files://` adopt this because a deleted file is also no longer valid and should result in the editor being closed (at least that is our behaviour today).

**Comment by jrieken** [comment]
Last time I talked to @joaomoreno we figured its the workbench not releasing the inputs but I could be wrong

**Comment by joaomoreno** [comment]
@jrieken The workbench doesn't know when it can release the inputs. There's a concept missing: the death of a Resource, regardless of how many people are holding on to its TextModel.

**Comment by jrieken** [comment]
Is this still an issue? Do inputs couple their lifecycle to editor model references? My understanding is that with the latest change @bpasero did the input just holds a reference and doesn't look at what models are alive

**Comment by bpasero** [comment]
I think the missing concept is still that a text content provider can signal a resource should be disposed because it is no longer valid. When I commit my changes I would expect any input that I opened from that commit to close (e.g. diff editors). Not sure though if this is a concept we need to add as API to extensions or rather the SCM viewlet being the one that closes any editor opened as soon as there are no more outgoing changes in a group.

**Comment by ProLoser** [comment]
FYI: I use an auto markdown preview extension which had a similar problem. Closing the source document that the preview pane is created and synchronized with doesn't close the corresponding preview. Additionally, opening multiple documents can sometimes mess up or shuffle tabs around. I'm not sure if it's relevant but it may be helpful to keep that in mind so that other developers are able to handle these problems more easily too.

https://github.com/hnw/vscode-auto-open-markdown-preview/issues/2
https://github.com/hnw/vscode-auto-open-markdown-preview/issues/7
https://github.com/hnw/vscode-auto-open-markdown-preview/issues/10

**Comment by jessejanderson** [comment]
Is there any progress on this? New user to VSCode and it's a little annoying and weird that I have to close these all the time. Even if it wasn't automatic I would love even just a `Close all working trees` command that would close them so I don't have to do it manually every time.

**Comment by dlong500** [comment]
To me at the very least it would seem appropriate to have all "working tree" editors grouped in their own tree in the explorer pane, possible with an option to hide that tree entirely unless you are in the "Source Control" view.  This would greatly reduce the clutter and grunt-work of closing those editors all the time.

**Comment by ericnewton76** [comment]
Also, its bizarre discarding changes made by my cat's paws on the keyboard from the night before, and still seeing the changes in the document after going to Source control pane and Discarding Changes on that file.  

Obviously the problem here is the cat trying to code and not knowing basic syntax and grammar, however, the Discard Changes feature I would intuitively expect it to also revert the changes in the opened document tab (or ask to revert?)  

Instead the changes are still visible in the document tab, even though the underlying file has been reverted.  You end up with a modified file confirmation box that could possibly reintroduce "speculative code" within files, and need to Discard Changes yet again on that file.

**Comment by OneOfOne** [comment]
I hate to be that guy, but any news?

**Comment by spacewaffle** [comment]
Still looking for something on this. Anything new?

**Comment by joaomoreno** [comment]
@jrieken I wonder if we're now closer to fix this?

**Comment by jrieken** [comment]
not really - you can be a file system and send a delete event but I think that will make us keep the editor input anyways

**Comment by dlong500** [comment]
Regardless of whether an editor is "stale" or not, is there a way to accomplish my earlier suggestion of at least grouping all these working tree/index editors within their own tree rather than scattered among the other editors?  I personally don't ever see a use for these editors unless I'm in the source control pane (at which point they are handled differently anyway).

**Comment by aapis** [comment]
Also hoping for a solution here in a future release. 👍

**Comment by rysport** [comment]
Seeing that this issue is now two years old, would it be possible to implement what @dlong500 suggested for now?

**Comment by maxcct** [comment]
Just going to throw an additional voice behind this request. Ideally I'd like to have a setting that would make working tree editors close themselves automatically as soon as another editor is switched to. Otherwise I always end up accumulating a silly quantity of open working tree editor tabs.

**Comment by jjhampton** [comment]
I'd love to see this request implemented as well. VSCode is the only code-editor I can think of that has this functionality - you inspect some files in the VCS editor to see what the diff is, but after you're done, there are multiple 'working tree' file views left open that have to be manually close.

**Comment by OneOfOne** [comment]
Almost 3 years later and no comments :-/

**Comment by rolandas-valantinas** [comment]
Over 3 years now!
This extremely annoying and prevents using opened tab limit - while checking for changes it closes actual development tabs!

**Comment by ataiya** [comment]
I add myself to the list of gripes.
In some editors code diffs are a (globa) "toggle" somewhere in the UI.
Keeping an open tab for them is crazy!

**Comment by rolandas-valantinas** [comment]
Can we just have option for diff to open in new window? So we at least could close all of them in one go!

**Comment by ataiya** [comment]
I think have diff tabs hide when you exit the "Source Control" mode (vs.
Editing, Search, ...) would likely be the best.
In fact perhaps all open files should have diff enabled in that mode?

**Comment by leventku** [comment]
I like the `enablePreview: false` mode but diff tabs getting too much accumulated. I think this should be managed better in vs code, either with this feature or some other way.

**Comment by sethrevelle** [comment]
Love VSCode, this is my biggest gripe though.  Workflow of double checking changes, before committing is made...... painful because of the lack of any way to close all working tree files.

**Comment by andreascian** [comment]
> I think have diff tabs hide when you exit the "Source Control" mode (vs. Editing, Search, ...) would likely be the best. In fact perhaps all open files should have diff enabled in that mode?

I got the same idea: there should be an option to close all diff tabs when going out of source control mode

**Comment by mholtzhausen** [comment]
> > I think have diff tabs hide when you exit the "Source Control" mode (vs. Editing, Search, ...) would likely be the best. In fact perhaps all open files should have diff enabled in that mode?
> 
> I got the same idea: there should be an option to close all diff tabs when going out of source control mode

Or close all diff views on commit

**Comment by phil294** [comment]
This issue would have been resolved with what was discussed in #85237: Adding a setting to always open working tree files in preview mode

the issue got closed by vscodebot though, so... whatever

**Comment by aapis** [comment]
Still would support this feature, but as a workaround I've stopped using the built-in git integration.  Works great!

**Comment by phil294** [comment]
> @phil294 what do you use instead?

@taiya I simply accepted my fate and set `workbench.editor.enablePreview` to `true` and learned to live with it, which I think is the only feasible way to use vscode with git integration

**Comment by andreascian** [comment]
> Still would support this feature, but as a workaround I've stopped using the built-in git integration. Works great!

[edamagit](https://github.com/kahole/edamagit) rules!! :+1:

**Comment by dlong500** [comment]
@joaomoreno  I hate to (also) be that guy, but this issue has been hanging around for far too long (4+ years).  Can we get some traction on at least a path forward with some options?  Honestly at this point I'd be happy with a setting that would simply hide any working tree editors unless you are in the source control view.  It just makes no sense whatsoever to have to manually close a working tree editor in the explorer view, and I haven't seen any reason presented why anyone would actually _want_ the current behavior.

**Comment by joaomoreno** [comment]
@dlong500 It's not that we don't care to fix it. But rather the large scope around the fix and the dependency on API, since git is an extension itself. This is blocked by https://github.com/microsoft/vscode/issues/15178.

**Comment by taiya** [comment]
Thanks for the diplomatic answer Joao, can we do anything on our side to speed up #15178?
This UI "bug" is one of the worst behaviors in the whole vscode IMHO.

Thanks!

**Comment by dlong500** [comment]
Thanks @joaomoreno.  I do want to clarify that I'm really not trying to be critical of vscode overall.  I love the app and appreciate all the amazing work that has been done these last many years by you and other developers.  This is just one of those things that I deal with literally every day so I would hope it would be taken off the backburner.  This is really the _only_ thing that truly bugs me about using vscode.  If I could contribute anything useful towards a resolution I'd certainly offer my limited skillset as well.

**Comment by omidantilong** [comment]
Adding another voice to this. Does it really need to be so complicated? It seems like a lot of time has been spent thinking about the context in which to show the diff editor, intelligently closing it etc. 

It doesn't have to be smart. Literally just something in the command palette that will close any tabs with `(Working Tree)` in their name would more than suffice at this point.

**Comment by lszomoru** [comment]
Today's Insiders release (2022-02-10) contains a new setting (`git.closeDiffOnOperation`) that is meant to address the need to manually close the diff editors after certain git operations. If the setting is enabled, diff editors will automatically be closed when changes are committed, discarded, staged, unstaged. Try out the new setting and let me know your feedback.

**Comment by taiya** [comment]
@lszomoru that's a good first step, but it would also be great if diff windows were limited to the "Source Control" tab?
(i.e. hidden and/or closed when you move to the other modes? explorer, search, ...)

**Comment by lszomoru** [comment]
Today's Insider build (2022-02-16) also contains a new command, `Git: Close All Diff Editors` that is available in the command palette and can be used to manually close all diff editors (both index, and working tree) independent of the value of the `git.closeDiffOnOperation` setting.

As for the request for the diff editors to only show when the SCM viewlet is active, I am most certain that its not currently possible. There is currently no support to associate an editor with a particular viewlet so that it is only being shown when the viewlet is active. This in scope is also greater than just SCM - imagine that you open the Extensions viewlet, search for an extension, and you click on it to show its details. It would be nice if the extension details editor would only show when the Extensions viewlet is active. I have filed [a separate issue](https://github.com/microsoft/vscode/issues/143193) to track adding support for associating an editor with a viewlet.

**Comment by lszomoru** [comment]
I am going to close this issue for now, while the concept of associating an editor with a viewlet is being discussed in [the linked issue](https://github.com/microsoft/vscode/issues/143193). If you are passionate about that, please add your feedback to that issue. In case there is a positive resolution of that issue, we can adopt that capability for the diff editors.

**Comment by dlong500** [comment]
Just a note that the newer related issue #143193 is now officially a candidate for the backlog of feature implementations, but it needs 20 upvotes to pass the threshold.  Please head over there and upvote if you haven't already.
